providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,300 @@
+
+    return ossl_hybrid_settable_ctx_params(ctx, provctx);
+}
+
+const OSSL_PARAM *ossl_hybrid_kem_gettable_ctx_params(void *vctx, void *provctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_gettable_ctx_params(ctx, provctx);
+}
+
+int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t total_len = ossl_hybrid_get_key_length_from_ctx(ctx);
+    size_t a, b, encsum = 0, secretsum = 0;
+    size_t elen[MAX_HYBRID_ALGS], slen[MAX_HYBRID_ALGS];
+    unsigned int i;
+
+    if (total_len == 0)
+        return 0;
+    /* Calculate lengths */
+    for (i = 0; i < ctx->num_algs; i++) {
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], NULL, i + elen,
+                                  NULL, i + slen))
+            return 0;
+        encsum += elen[i];
+        secretsum += slen[i];
+    }
+    if (enc == NULL) {
+        if (enclen != NULL)
+            *enclen = encsum;
+        if (secretlen != NULL)
+            *secretlen = secretsum;
+        return enclen != NULL || secretlen != NULL;
+    }
+
+    /* Error checking */
+    if (*secretlen < secretsum) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*secretlen too small");
+        return 0;
+    }
+    if (*enclen < encsum) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*enclen too small");
+        return 0;
+    }
+
+    /* Set outputs */
+    *enclen = encsum;
+    *secretlen = secretsum;
+
+    /* Encapsulate */
+    for (i = 0; i < ctx->num_algs; i++) {
+        a = elen[i];
+        b = slen[i];
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &a, secret, &b))
+            return 0;
+        if (a != elen[i] || b != slen[i]) {
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                           "secretlen or enclen changed");
+            return 0;
+        }
+        secret += slen[i];
+        enc += elen[i];
+    }
+    return 1;
+}
+
+int ossl_hybrid_kem_decapsulate(void *vctx,
+                                unsigned char *secret, size_t *secretlen,
+                                const unsigned char *enc, size_t enclen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t total_len = ossl_hybrid_get_key_length_from_ctx(ctx);
+    size_t a, encsum = 0, secretsum = 0;
+    size_t elen[MAX_HYBRID_ALGS], slen[MAX_HYBRID_ALGS];
+    unsigned int i;
+
+    if (total_len == 0)
+        return 0;
+    /* Calculate lengths */
+    for (i = 0; i < ctx->num_algs; i++) {
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], NULL, i + slen,
+                                  NULL, i + elen))
+            return 0;
+        encsum += elen[i];
+        secretsum += slen[i];
+    }
+    if (secret == NULL) {
+        *secretlen = secretsum;
+        return 1;
+    }
+    if (*secretlen < secretsum) {
```

Comment:
Assumption is secretlen cant be NULL.

Reply:
Yes.  You won't know the output length otherwise.

Reply:
Is this a problem?

The size needs to be passed in and returned?  It doesn't make sense for it to be NULL.

Reply:
This is the [normal pattern for all these APIs](https://docs.openssl.org/master/man3/EVP_PKEY_decapsulate/#examples), no? Pass in NULL data (`secret` in this case) and use `secretlen` to find out actual length. Rince-repeat (2 times in this case). fwiw, lgtm.